src/index/Index.Text.cpp

```diff
@@ -397,14 +397,24 @@ void Index::calculateBlockBoundaries() {
   // 4) word.substring(0, MIN_PREFIX_LENGTH) is different from the next.
   // A block boundary is always the last WordId in the block.
   // this way std::lower_bound will point to the correct bracket.
+  if (_textVocab.size() == 0) {
+    LOG(WARN) << "You are trying to call calculateBlockBoundaries on an empty "
+                 "text vocabulary\n";
+    return;
+  }
+  const auto& locManager = _textVocab.getLocaleManager();
+  auto currentLenAndPrefix = LocaleManager::getUTF8Prefix(
+      _textVocab[0].value().get(), MIN_WORD_PREFIX_SIZE);
   for (size_t i = 0; i < _textVocab.size() - 1; ++i) {
     // we need foo.value().get() because the vocab returns
     // a std::optional<std::reference_wrapper<string>> and the "." currently
     // doesn't implicitly convert to a true reference (unlike function calls)
-    if (_textVocab[i].value().get().size() < MIN_WORD_PREFIX_SIZE ||
-        (_textVocab[i + 1].value().get().size() < MIN_WORD_PREFIX_SIZE) ||
-        _textVocab[i].value().get().substr(0, MIN_WORD_PREFIX_SIZE) !=
-            _textVocab[i + 1].value().get().substr(0, MIN_WORD_PREFIX_SIZE)) {
+    auto nextLenAndPrefix = LocaleManager::getUTF8Prefix(
+        _textVocab[i + 1].value().get(), MIN_WORD_PREFIX_SIZE);
+    if (currentLenAndPrefix.first < MIN_WORD_PREFIX_SIZE ||
+        (nextLenAndPrefix.first < MIN_WORD_PREFIX_SIZE) ||
+        locManager.compare(currentLenAndPrefix.second, nextLenAndPrefix.second,
+                           LocaleManager::Level::PRIMARY) > 0) {
```

Comment:
Ok, this is a little bit tricky. Everywhere, where we answer the question "is `a` a prefix of `b` we are always comparing
on the primary level because I've found no way to do this on any other Level using ICU. I am thinking about writing those functions for ICU, but this requires getting to know their codebase first.

In the Filter this is a design decision if we want to have exact maches, or case-insensitive matches etc. To me it is not super clear what we want there so I am open for suggestions.

Reply:
The ICU documentation says:

 "Identical Level: When all other levels are equal, the identical level is used as a tiebreaker. The Unicode code point values of the NFD form of each string are compared at this level, just in case there is no difference at levels 1-4 . For example, Hebrew cantillation marks are only distinguished at this level. This level should be used sparingly, as only code point values differences between two strings is an extremely rare occurrence. Using this level substantially decreases the performance for
both incremental comparison and sort key generation (as well as increasing the sort key length). It is also known as level 5 strength."

So mayve we should not use that in Filters which are actually more performance critical than the binary searches. I think from a user perspective missing something unexpectedly in a Filter is often worse than getting more matches.

Reply:
I switched all the `IDENTICAL`S to `QUARTERNARY`s.
But, as I think about it now: The Filters also only use binary searches to obtain the legal ID ranges for the results, so I don't expect this to make much of a difference. (The only mechanism that actually looks at a lot of strings is the regex filter which is exactly because of this reason typically infeasible.